lxd/storage/drivers/drivers_lvm.go

```diff
@@ -0,0 +1,561 @@
+		if err == nil {
+			if lvCount == 0 {
+				removeVg = true // Volume group is totally empty, safe to remove.
+			} else if d.usesThinpool() && lvCount > 0 {
+				// Lets see if the lv count is just our thin pool, or whether we can only remove
+				// the thin pool itself and not the volume group.
+				thinVolCount, err := d.countThinVolumes(d.config["lvm.vg_name"], d.thinpoolName())
+				if err != nil && err != errLVMNotFound {
+					return err
+				}
+
+				// Thin pool exists.
+				if err == nil {
+					// If thin pool is empty and the total VG volume count is 1 (our thin pool
+					// volume) then just remote the entire volume group.
+					if thinVolCount == 0 && lvCount == 1 {
+						removeVg = true
+					} else if thinVolCount == 0 && lvCount > 1 {
+						// Otherwise, if the thin pool is empty but the volume group has
+						// other volumes, then just remove the thin pool volume.
+						err = d.removeLogicalVolume(d.lvmDevPath(d.config["lvm.vg_name"], "", "", d.thinpoolName()))
+						if err != nil {
+							return fmt.Errorf("Failed to delete thin pool \"%s\" from volume group \"%s\": %v", d.thinpoolName(), d.config["lvm.vg_name"], err)
+						}
+						d.logger.Debug("Thin pool removed", log.Ctx{"vg_name": d.config["lvm.vg_name"], "thinpool_name": d.thinpoolName()})
+					}
+				}
+			}
+		}
+
+		// Remove volume group if needed.
+		if removeVg {
+			_, err := shared.TryRunCommand("vgremove", "-f", d.config["lvm.vg_name"])
+			if err != nil {
+				return fmt.Errorf("Failed to delete the volume group for the lvm storage pool: %v", err)
+			}
+			d.logger.Debug("Volume group removed", log.Ctx{"vg_name": d.config["lvm.vg_name"]})
+		} else {
+			// Otherwise just remove the lvmVgPoolMarker tag to indicate LXD no longer uses this VG.
+			if shared.StringInSlice(lvmVgPoolMarker, vgTags) {
+				_, err = shared.TryRunCommand("vgchange", "--deltag", lvmVgPoolMarker, d.config["lvm.vg_name"])
+				if err != nil {
+					return fmt.Errorf("Failed to remove marker tag on volume group for the lvm storage pool: %v", err)
+				}
+				d.logger.Debug("LXD marker tag removed from volume group", log.Ctx{"vg_name": d.config["lvm.vg_name"]})
+			}
+		}
+	}
+
+	// If we have removed the volume group and this is a loop file, lets clean up the physical volume too.
+	if removeVg && loopFile != nil {
+		_, err := shared.TryRunCommand("pvremove", "-f", loopFile.Name())
+		if err != nil {
+			d.logger.Warn("Failed to destroy the physical volume for the lvm storage pool", log.Ctx{"err": err})
+		}
+		d.logger.Debug("Physical volume removed", log.Ctx{"pv_name": loopFile.Name()})
+
+		// Set LO_FLAGS_AUTOCLEAR before removing the loop file otherwise we will get EBADF.
+		err = SetAutoclearOnLoopDev(int(loopFile.Fd()))
+		if err != nil {
+			d.logger.Warn("Failed to set LO_FLAGS_AUTOCLEAR on loop device, manual cleanup needed", log.Ctx{"err": err})
+		}
+
+		err = loopFile.Close()
+		if err != nil {
+			return err
+		}
```

Comment:
This can go away and be replaced by a call to `Unmount()` during the Delete logic. At which point all we need to do is delete the disk file if present, similar to what we do in btrfs.